src/codemagic/firebase/resource_managers/resource_manager.py

```diff
@@ -0,0 +1,105 @@
+from abc import ABC
+from abc import abstractmethod
+from enum import Enum
+from typing import Any
+from typing import Dict
+from typing import Generic
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Type
+from typing import TypeVar
+
+from googleapiclient import discovery
+from googleapiclient.errors import Error as GoogleApiClientError
+from httplib2.error import HttpLib2Error
+from oauth2client.client import Error as OAuth2ClientError
+from typing_extensions import Protocol
+
+from ..api_error import FirebaseApiClientError
+from ..resources.resource import Resource
+
+ResourceT = TypeVar('ResourceT', bound=Resource)
+ParentResourceIdentifierT = TypeVar('ParentResourceIdentifierT', bound='ParentResourceIdentifier')
+
+
+class ParentResourceIdentifier(ABC):
+    @property
+    @abstractmethod
+    def uri(self) -> str:
+        ...
+
+
+class PExecutableInterface(Protocol):
+    def execute(self) -> Dict[str, Any]:
+        ...
+
+
+class PListableInterface(Protocol):
+    def list(self, **kwargs) -> PExecutableInterface:
+        ...
+
+
+class ResourceManager(Generic[ResourceT, ParentResourceIdentifierT], ABC):
```

Comment:
In general see [this](https://github.com/codemagic-ci-cd/cli-tools/compare/feature/firebase-releases...feature/firebase-releases-interface-modifications?expand=1). I don't think it is possible to properly make a generic resource manager with listing capabilities as there are resources which cannot be listed (for example [`projects.apps`](https://firebase.google.com/docs/reference/app-distribution/rest/v1/projects.apps)).